### DISCUSSION TITLE
Are there examples of drawing lines, circles and rectangles on the ssd1306 display?

### DISCUSSION THREAD

**Discussion by marklacas** [tabemann/zeptoforth#36] [Q&A]
I'm getting close but could use some examples.

**Comment by tabemann** [comment]
There is currently code for drawing lines and filled-in/copied rectangles, and rectangular edges can be drawn with lines.

For drawing lines there is the words in the module `bitmap-utils` in `extra/common/bitmap_utils.fs`:

`draw-pixel-line` ( const x0 y0 x1 y1 op dst -- )

This draws a line on bitmap *dst* with a constant single-pixel pen.

`draw-rect-line` ( const width height x0 y0 x1 y1 op dst -- )

This draws a line on bitmap *dst* with a constant rectangular pen.

`draw-bitmap-line` ( src-x src-y src-width src-height x0 y0 x1 y1 op src dst -- )

This draws a line on bitmap *dst* with a rectangular pen copied from another bitmap *src*.

Here *op* is a single-pixel, rectangle, or bitmap-to-bitmap operation, as defined in the module `bitmap` in `extra/common/bitmap.fs`; e.g. `draw-pixel-line` may take `['] xor-pixel-const`, `draw-rect-line` may take `['] or-rect-const`, and `draw-bitmap-line` may take `['] bic-rect-const`.

*const* is a 8-bit vertical bit pattern. In most cases you will want `$FF` for this.

*x0*, *y0*, *x1*, and *y1* should be self-explanatory.

*width*/*src-width* and *height*/*src-height* are the width and height of the rectangular pen. Note that the rectangular pen is centered on the line from (*x0*, *y0*) to (*x1*, *y1*).

*src-x* and *src-y* are the upper left-hand corner of the portion of the source bitmap used as a pen.

For drawing filled-in/copied there is the words in the module `bitmap` in `extra/common/bitmap.fs`:

For both of these sets of words there are the following operations: `set` (set pixels), `or` (or pixels), `and` (and pixels), `bic` (clear pixels), and `xor` (exclusive-or pixels). The two sets of words are:

*op*`-rect-const` ( const start-dst-col col-count start-dst-row row-count dst -- )

Draw a solid rectangle on bitmap *dst* with the specified operation *op*. *const* is as above.

*op*`-rect` ( start-src-col src-dst-col col-count start-src-row start-dst-row row-count src dst -- )

Apply a rectangle within bitmap *src* to bitmap *dst* with the specified operation *op*.

*start-dst-col* and *start-dst-row* are the starting column and row of the rectangle in bitmap *dst*.

*col-count* and *row-count* are the column and row count of the rectangle.

*start-src-col* and *start-src-row* are the starting column and row of the rectangle in bitmap *src*.

**Reply by tabemann** [reply]
I should note, however, that there is currently no code for drawing circles, ellipses, or arcs. If you write any code to that effect, I'd be happy to add it to zeptoforth.

**Reply by tabemann** [reply]
I should also note that these words work for all instances of `<bitmap>`, which includes but is not limited to instances of `<ssd1306>`, so you can use them both on SSD1306 displays and on offline bitmaps.

**Comment by marklacas** [comment]
Yes, I'm aware of all those calls, but not quite clear on the usage with the SSD1306 displays.
Could you provide one quick use case? Maybe how to draw a line or rectangle.   After that I will be good to go and will extend as I can.
You are welcome to anything that I come up with.

**Reply by tabemann** [reply]
Here is some example code:

```
#include extra/common/bitmap.fs
#include extra/common/bitmap_utils.fs
#include extra/common/ssd1306.fs

oo import
bitmap import
bitmap-utils import
ssd1306 import

\ We are using a 128x64 display; update this to 128x32 if you so see fit.
128 constant my-width
64 constant my-height

\ Here we create the SSD1306 display and its backing buffer.
my-width my-height bitmap-buf-size constant my-buf-size
my-buf-size 4 align buffer: my-buf
<ssd1306> class-size buffer: my-ssd1306
14 15 my-buf my-width my-height SSD1306_I2C_ADDR 1 <ssd1306> my-ssd1306 init-object

\ We are defining a 6x6 "brush" bitmap to draw to the SSD1306 device with.
6 constant my-brush-width
6 constant my-brush-height

\ Its definition is similar to that of an SSD1306 device except it does not map to any GPIO pins,
\ have an I2C address, or use an I2C peripheral.
my-brush-width my-brush-height bitmap-buf-size constant my-brush-buf-size
my-brush-buf-size 4 align buffer: my-brush-buf
<bitmap> class-size buffer: my-brush
my-brush-buf my-brush-width my-brush-height <bitmap> my-brush init-object

\ Here we draw a small circle to the "brush" bitmap.
$FF 2 2 0 my-brush-height my-brush set-rect-const
$FF 0 my-brush-width 2 2 my-brush set-rect-const
$FF 1 4 1 4 my-brush set-rect-const

\ Here we draw a single pixel-wide line from the upper left-hand corner to the lower
\ right-hand corner of the display.
$FF 0 0 my-width 1- my-height 1- ' set-pixel-const my-ssd1306 draw-pixel-line

\ Here we draw a line with a 4x4 rectangular pen from the lower left-hand corner to the upper
\ right-hand corner of the display.
$FF 4 4 0 my-height 1- my-width 1- 0 ' set-rect-const my-ssd1306 draw-rect-line

\ Here we draw a line with the "brush" bitmap as a pen through the middle horizontal third of the display.
0 0 my-brush-width my-brush-height my-width 3 / my-height 2 / my-width 3 / 2 * my-height 2 /
' or-rect my-brush my-ssd1306 draw-bitmap-line

\ Here we exclusive-or a rectangle in the middle of the display.
$FF my-width 4 / my-width 2 / my-height 4 / my-height 2 / my-ssd1306 xor-rect-const

\ Here we exclusive-or a single instance of the "brush" bitmap in the middle of the display.
0 my-width 2 / my-brush-width 2 / - my-brush-width
0 my-height 2 / my-brush-height 2 / - my-brush-height
my-brush my-ssd1306 xor-rect

\ Here we update the display.
my-ssd1306 update-display
```

One subtlety I should note is that you should not use `xor-rect-const` with `draw-rect-line` or `xor-rect` or `set-rect` with `draw-bitmap-line` because they will very likely have undesirable results. I would suggest using either `or-rect-const` or `bic-rect-const` with `draw-rect-line` or `or-rect` or `bic-rect` with `draw-bitmap-line`.

Here is an image containing the results of the code above (mind you where the pixel line crosses the big rectangle does not show up due to the limitations of my phone's camera):

![20230313_214745](https://user-images.githubusercontent.com/4637032/224880899-cd1c9117-49c4-4825-9998-7b7b4bc5889e.jpg)

**Reply by marklacas** [reply]
Perfect! That's what I needed.  Now to expand it with arcs and circles.
Thanks!

**Reply by marklacas** [reply]
I've got circles and filled circles working in zeptoforth. Also made a battery word.
![IMG_0124](https://user-images.githubusercontent.com/566089/225662610-dbc07425-b4c0-425f-96f2-8c39ef37d412.jpeg)

**Reply by marklacas** [reply]
I'll provide the code when I'm finished cleaning it up.

**Reply by tabemann** [reply]
Nice! I would really like to integrate these into zeptoforth once they're ready.
